src/OpenApi/JSON.hs

```diff
@@ -0,0 +1,89 @@
+module OpenApi.JSON
+  ( generateRenamed
+  , parseJSONFixed
+  , parseRefSum
+  , parseRenamed
+  )
+where
+
+import Data.Map.Strict (Map)
+import Data.String (String)
+import GHC.Enum (Bounded, Enum, minBound)
+import GHC.Generics (Generic, Rep)
+import OpenApi.Prelude
+
+import qualified Data.Aeson.Types    as JSON
+import qualified Data.HashMap.Strict as HashMap
+import qualified Data.List           as List
+import qualified Data.Map.Strict     as Map
+import qualified Data.Text           as Text
+
+parseRenamed
+  :: (Generic a, JSON.GFromJSON JSON.Zero (Rep a))
+  => Map String String
+  -> JSON.Value
+  -> JSON.Parser a
+parseRenamed = JSON.genericParseJSON . renameOptions
+
+generateRenamed
+  :: (Generic a, JSON.GToJSON JSON.Zero (Rep a))
+  => Map String String
+  -> a
+  -> JSON.Value
+generateRenamed = JSON.genericToJSON . renameOptions
+
+parseJSONFixed
+  :: forall a b c . (Bounded b, Enum b, Eq a, Show a)
+  => Text
+  -> (String -> (a -> JSON.Parser b) -> c -> JSON.Parser b)
+  -> (b -> a)
+  -> c
+  -> JSON.Parser b
+parseJSONFixed name withValue map =
+  withValue (convertText name) $ \value ->
+    maybe (unexpected value) pure $ List.lookup value options
+  where
+    options :: [(a, b)]
+    options = (\item -> (map item, item)) <$> [minBound..]
+
+    unexpected :: a -> JSON.Parser b
+    unexpected value
+      = fail
+      . convertText
+      $ "Unexpected " <> name <> ": " <> convertText (show value)
+
+parseRefSum
+  :: forall a b c . (JSON.FromJSON a)
+  => (Text -> b)
+  -> (a -> c)
+  -> (b -> c)
+  -> Text
+  -> Text
+  -> JSON.Value
+  -> JSON.Parser c
+parseRefSum mkName mkBody mkRef prefix name value =
```

Comment:
OT, I wonder about how many formal function parameters would typically be considered a smell in HS.

Reply:
To me as long each parameter has a unique type: There is no upper limit.

Reply:
> To me as long each parameter has a unique type: There is no upper limit.

I think there probably is an upper limit in terms of complexity. If a function takes 20 parameters, there's an excellent chance that it could be further decomposed into smaller functions.

I'm not saying this particular function does too much, because I haven't looked at it recently and don't recall what it does, but I think the more parameters the more closely we should look at it to see if it can be decomposed.